src/rpc/server.zig

```diff
@@ -0,0 +1,554 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+
+const SnapshotGenerationInfo = sig.accounts_db.AccountsDB.SnapshotGenerationInfo;
+const FullSnapshotFileInfo = sig.accounts_db.snapshots.FullSnapshotFileInfo;
+const IncrementalSnapshotFileInfo = sig.accounts_db.snapshots.IncrementalSnapshotFileInfo;
+const ThreadPool = sig.sync.ThreadPool;
+
+const logger_scope = "rpc.Server";
+const ScopedLogger = sig.trace.log.ScopedLogger(logger_scope);
+
+pub const Server = struct {
+    //! Basic usage:
+    //! ```zig
+    //! var server = try Server.init(.{...});
+    //! defer server.joinDeinit();
+    //!
+    //! try server.serveSpawnDetached(); // or `.serveDirect`, if the caller can block or is managing the separate thread themselves.
+    //! ```
+
+    allocator: std.mem.Allocator,
+    logger: ScopedLogger,
+
+    snapshot_dir: std.fs.Dir,
+    latest_snapshot_gen_info: *sig.sync.RwMux(?SnapshotGenerationInfo),
+
+    /// Wait group for all currently running tasks, used to wait for
+    /// all of them to finish before deinitializing.
+    wait_group: std.Thread.WaitGroup,
+    thread_pool: *ThreadPool,
+
+    /// Must not be mutated.
+    read_buffer_size: usize,
+    tcp: std.net.Server,
+
+    pub const InitParams = struct {
+        /// Must be a thread-safe allocator.
+        allocator: std.mem.Allocator,
+        logger: sig.trace.Logger,
+
+        /// Not closed by the `Server`, but must live at least as long as it.
+        snapshot_dir: std.fs.Dir,
+        /// Should reflect the latest generated snapshot eligible for propagation at any
+        /// given time with respect to the contents of the specified `snapshot_dir`.
+        latest_snapshot_gen_info: *sig.sync.RwMux(?SnapshotGenerationInfo),
+
+        thread_pool: *ThreadPool,
+
+        /// The size for the read buffer allocated to every request.
+        /// Clamped to be greater than or equal to `MIN_READ_BUFFER_SIZE`.
+        read_buffer_size: u32,
+        /// The socket address to listen on for incoming HTTP and/or RPC requests.
+        socket_addr: std.net.Address,
+
+        pub const MIN_READ_BUFFER_SIZE = 256;
+    };
+
+    pub const InitError = std.net.Address.ListenError;
```

Comment:
I don't think we need this

Reply:
Forgot to use this as the error set for `init`.

Reply:
4b7beebaff498199724eeb771ca4cd4215f015d8

Reply:
I sort of think that given it is only a single error, it is clearer to not alias it. lmk your thoughts

Reply:
Well, it isn't a single error, it's an error set, specifically a union of 5 error sets.

Reply:
pub const InitError = std.net.Address.ListenError;

Reply:
is that not a single error?

Reply:
I think this is great
```zig
const AcceptConnectionError = error{
    ProcessFdQuotaExceeded,
    SystemFdQuotaExceeded,
    SystemResources,
    ProtocolFailure,
    BlockedByFirewall,
    NetworkSubsystemFailed,
} || std.posix.UnexpectedError;
```

Reply:
Further, this is a public API, and I would posit that all public APIs should ideally describe their error set.

Reply:
I think this is going a little far and is overly verbose, would you also alias 
```zig
pub fn base58EncodeAlloc(self: Hash, allocator: Allocator) Allocator.Error![]const u8 {
    return base58.encodeAlloc(self.data, allocator);
}
```
to 
```zig
const Base58EncodeError = Allocator.Error;
pub fn base58EncodeAlloc(self: Hash, allocator: Allocator) Base58EncodeError![]const u8 {
    return base58.encodeAlloc(self.data, allocator);
}
```

Reply:
I think allocation errors are a particular case of convention being sufficient; everyone intuitively understands that if you pass in an allocator, and an error is returned, among those errors will be `std.mem.Allocator.Error`, and can easily reach for it; in this case, and most other similar cases, it is not obvious that `init` would return `std.net.Address.ListenError`.

Reply:
I agree with harold on this one tbh - lets not over alias things which dont need to be

Reply:
3761d42d41ee276fdc40dfb74e0499d34ac74249